src: separate trace_event_common from trace_event

targos opened the pull request:
Currently, trace_event.h contains a full copy of the header present in
deps/v8/base/trace_event/common/trace_event_common.h mixed with
implementation-specific macros.
This patch moves the V8 code to its own file and includes it from
trace_event.h.

<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x]  `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] commit message follows commit guidelines

<!-- Provide affected core subsystem(s) (like doc, cluster, crypto, etc). -->

src

targos commented:
/cc @jasongin @nodejs/diagnostics @nodejs/v8 

I'm happy to remove the copy and directly include it from `deps` if you want.

jasongin commented:
LGTM, certainly an improvement, though I don't know if it's the *best* approach.

@ofrobots do you know why `trace_event_common.h` was originally added at [`deps\v8\trace_event\common`](https://github.com/nodejs/node/tree/master/deps/v8/base/trace_event/common) rather than [`deps\v8\include`](https://github.com/nodejs/node/tree/master/deps/v8/include)? If it was in the latter location then it would be easier to just `#include` in `trace_event.h` here.

Ideally @matthewloring should also comment on this, but I don't know when he is going to be back at work.

ofrobots commented:
IIRC it was quite intentional that @matthewloring did not directly include `trace_event_common.h` – but the details escape me. Perhaps @fmeawad knows too. Matt should be back mid next week, he has the full context here.

Perhaps we can go with https://github.com/nodejs/node/pull/10623 as the workaround for now and figure out a longer-term solution (this PR or otherwise) once @matthewloring is back.

jasongin commented:
https://github.com/nodejs/node/pull/10623 is not a workaround for this issue, it is just a fix for a separate tracing-related bug.

ofrobots commented:
Ah sorry, I misunderstood. In that case the work-around might be to copy over the necessary changes from `trace_event_common.h` in order to unblock https://github.com/nodejs/node/pull/9618? @targos, I can put that together if you want, but it might be simpler for you to add the changes in #9618.

matthewloring commented:
This change is definitely a step in the right direction. The reason I did not depend on the file directly was to avoid a dependency on a directory outside of `deps/v8/include`. IIRC the file was put outside of `deps/v8/include` because it was originally copied from Chromium but @fmeawad would have more context.

joshgav commented:
V8 includes trace_event_common.h [here](https://github.com/v8/v8/blob/790518c13043ba0c83fa7cf9c240ed9f151bd1e3/src/tracing/trace-event.h#L11) with `#include "base/trace_event/common/trace_event_common.h"`, which relies on the upstream `src/base/trace_event/common` repo in Chromium [here](https://chromium.googlesource.com/chromium/src/base/trace_event/common/).

@ofrobots @fmeawad would it be a good idea for Node to include the "trace_event_common" repo from Chromium directly under `deps` here in Node, rather than copy the file to `deps/v8` as currently? And long term should we rely on one of these upstreams or our own fork as in #9304 and here?

joshgav commented:
Conclusion in discussion in Diag WG meeting today is that Node should manage its copy of the trace_event headers however we see fit and not worry about upstream Chromium. V8/Google plan to provide a proper tracing library eventually and when that's ready we can sync with it upstream.

targos commented:
@jashgav thanks. Does it mean that we should go on with this PR?

targos commented:
Rebased and updated to the latest version from V8 5.6. `src/tracing/trace_event_common.h` is now a copy of `deps/v8/base/trace_event/common/trace_event_common.h`.

jasnell commented:
Any updates on this one?

targos commented:
I'm just waiting for some reviews

bnoordhuis commented on src/tracing/trace_event_common.h:
```diff
@@ -0,0 +1,1073 @@
+// Copyright 2015 The Chromium Authors. All rights reserved.
```
This file doesn't need include guards?

targos commented on src/tracing/trace_event_common.h:
```diff
@@ -0,0 +1,1073 @@
+
+#if defined(TRACE_EVENT0)
+#error "Another copy of this file has already been included."
+#endif
```
@bnoordhuis It has this ^

bnoordhuis replied:
Right, but the file you moved that from has traditional include guards.  I wondered (and wonder) if there is a reason not to add them here.

targos replied:
I didn't think about that. This is just a copy of the file from V8. Would you like me to add include guards?

joshgav commented:
LGTM.

But it doesn't seem to fix the break noted [here (comment)](https://github.com/nodejs/diagnostics/issues/91#issuecomment-288135669), per my testing with [this follow-on commit](https://github.com/joshgav/node/commit/b10b8ea75bd5cbd32a4a7a26358fd9c89f6e829c).

cc @matthewloring

matthewloring commented:
There are likely more changes required to account for updates to the macros in V8 5.7 which just landed on master. I'm currently working to put together a set of steps for updating these macros when V8 is updated. I also have a patch that will fix these files for 5.7. I can post the steps I've been using here or open the patch separately. @targos What would you prefer?

targos commented:
@matthewloring go ahead and open a PR

matthewloring commented:
Opened here: #12127

targos commented:
Thanks. I'll close this one then.
